subcommands/version/updates.go

```diff
@@ -91,6 +99,72 @@ func (f *FioctlUpdateFinder) FindLatest() (*FioctlUpdate, error) {
 	return update, nil
 }
 
+type progressBar struct {
+	total   int64
+	written int64
+	sha     hash.Hash
+	buff    bytes.Buffer
+}
+
+func (p *progressBar) Write(buff []byte) (n int, err error) {
+	if _, err = p.sha.Write(buff); err != nil {
+		return 0, err
+	}
+	n, err = p.buff.Write(buff)
+	p.written += int64(n)
+	// Print 20 dashes total
+	dashes := 100 * p.written / p.total / 5
+	spaces := 20 - dashes
+	progress := strings.Repeat("*", int(dashes)) + strings.Repeat(" ", int(spaces))
+	fmt.Printf("[%s] %d%%\r", progress, dashes*5)
+	return
+}
+
+func (u FioctlUpdate) Do() error {
+	exe, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("unable to find path to self: %w", err)
+	}
+	st, err := os.Stat(exe)
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("Path to self is: %s", exe)
+	fmt.Println("Downloading update:", u.Uri)
+	res, err := http.Get(u.Uri)
+	if err != nil {
+		return err
+	}
+	if res.ContentLength != u.len {
+		return fmt.Errorf("target size mismatch: %d != %d", res.ContentLength, u.len)
+	}
+	defer res.Body.Close()
+	pb := &progressBar{
+		total: u.len,
+		sha:   sha512.New(),
+		buff:  bytes.Buffer{},
+	}
+	if _, err := io.Copy(pb, io.LimitReader(res.Body, u.len)); err != nil {
+		return fmt.Errorf("unable to read response. HTTP_%d: %w", res.StatusCode, err)
+	}
+	fmt.Println()
+	if res.StatusCode != 200 {
+		return fmt.Errorf("unable to download %s. HTTP_%d: %s", u.Uri, res.StatusCode, pb.buff.String())
+	}
+	fmt.Println("Validating the checksum is", u.Sha512)
+	sha := pb.sha.Sum(nil)
```

Comment:
I think that hasher generates a hash/checksum as data goes through it, it doesn't need to store the overall file to generate checksum.
If so, then Volodya's comment makes sense, since we can stream down to the the tmp file.

Maybe it's possible to construct `socket-buffer --> limit reader --> hasher --> tmp.file` pipe that `io.Copy` would pipe data through.

Reply:
golang has a teewriter that can write to the hasher and file at the same time.